File: swift/Sources/GridMD/GridMD.swift
// Public API for the GridMD Swift library.
//
//   let json = try GridMD.dump(source)                 // canonical model dump
//   let xlsx = try GridMD.exportXLSX(source)            // GridMD → .xlsx
//   let back = try GridMD.importXLSX(xlsx.data)         // .xlsx → GridMD
//   let lint = GridMD.lint(source)                      // errors + warnings
//
// The library is pure Swift + Foundation (+ the system Compression framework);
// no third-party dependencies.

import Foundation

public enum GridMD {
    public struct Diagnostic: Equatable, Sendable {
        public let line: Int
        public let message: String
    }

    public struct LintResult: Sendable {
        public let errors: [Diagnostic]
        public let warnings: [Diagnostic]
        public let sheets: Int
        public let definedCells: Int
        public let blocks: Int
        public var isValid: Bool { errors.isEmpty }
    }

    public struct FidelityLine: Sendable {
        public let line: Int
        public let action: String
        public let feature: String
        public let note: String?
    }

    public struct ExportResult: Sendable {
        public let data: Data
        public let report: [FidelityLine]
    }

    public struct ImportResult: Sendable {
        public let gmd: String
        public let report: [FidelityLine]
    }

    public enum Failure: Error, CustomStringConvertible {
        case invalid([Diagnostic])
        case badXLSX(String)

        public var description: String {
            switch self {
            case let .invalid(diags): return "invalid GridMD: \(diags.count) error(s)"
            case let .badXLSX(msg): return "bad .xlsx: \(msg)"
            }
        }
    }

    /// Lints `source` (strict by default). Never throws — inspect `errors`.
    public static func lint(_ source: String, strict: Bool = true) -> LintResult {
        let l = runLint(source, mode: strict ? "strict" : "lenient")
        return LintResult(
            errors: l.errors.map { Diagnostic(line: $0.line, message: $0.msg) },
            warnings: l.warnings.map { Diagnostic(line: $0.line, message: $0.msg) },
            sheets: l.doc.sheets.count,
            definedCells: l.doc.statsDefs,
            blocks: l.doc.statsBlocks
        )
    }

    /// Canonical conformance dump. Throws `Failure.invalid` if strict lint fails.
    public static func dump(_ source: String) throws -> String {
        let l = runLint(source, mode: "strict")
        if !l.errors.isEmpty { throw Failure.invalid(l.errors.map { Diagnostic(line: $0.line, message: $0.msg) }) }
        return Dump.model(Model.build(l.doc))
    }

    /// Exports GridMD → `.xlsx` (worksheet core native + full-source carry part).
    /// Throws `Failure.invalid` if strict lint fails.
    public static func exportXLSX(_ source: String) throws -> ExportResult {
        let l = runLint(source, mode: "strict")
        if !l.errors.isEmpty { throw Failure.invalid(l.errors.map { Diagnostic(line: $0.line, message: $0.msg) }) }
        let (bytes, report) = Xlsx.write(model: Model.build(l.doc), source: source)
        return ExportResult(data: Data(bytes), report: report.map(mapReport))
    }

    /// Imports `.xlsx` → GridMD (restored losslessly from the carry part).
    public static func importXLSX(_ data: Data) throws -> ImportResult {
        do {
            let (gmd, report) = try Xlsx.read(Array(data))
            return ImportResult(gmd: gmd, report: report.map(mapReport))
        } catch {
            throw Failure.badXLSX(String(describing: error))
        }
    }

    private static func mapReport(_ r: ReportLine) -> FidelityLine {
        FidelityLine(line: r.line, action: r.action, feature: r.feature, note: r.note)
    }
}

struct InternalLint {
    let doc: Document
    let errors: [Diagnostic]
    let warnings: [Diagnostic]
}

/// Parses + validates, returning diagnostics sorted stably by line (matching the
/// JS reference's `[...].sort((a,b) => a.line - b.line)`).
func runLint(_ source: String, mode: String) -> InternalLint {
    let doc = Parser.parseDocument(source, mode: mode)
    validateDocument(doc)
    return InternalLint(doc: doc, errors: stableSortByLine(doc.errors), warnings: stableSortByLine(doc.warnings))
}

private func stableSortByLine(_ diags: [Diagnostic]) -> [Diagnostic] {
    diags.enumerated()
        .sorted { a, b in a.element.line != b.element.line ? a.element.line < b.element.line : a.offset < b.offset }
        .map { $0.element }
}

File: go/internal/gridmd/gridmd.go
// Package gridmd ties parsing and validation together (mirrors js/src/index.js
// `lint`) and exposes the canonical dump entry point.
package gridmd

import (
	"sort"

	"github.com/fledgling-co/gridmd/go/internal/dump"
	"github.com/fledgling-co/gridmd/go/internal/model"
	"github.com/fledgling-co/gridmd/go/internal/parser"
	"github.com/fledgling-co/gridmd/go/internal/validate"
)

// Result is a lint outcome: the parsed document plus line-sorted diagnostics.
type Result struct {
	Doc      *parser.Document
	Errors   []parser.Diag
	Warnings []parser.Diag
}

// Lint parses and validates source in strict mode.
func Lint(source string) *Result {
	doc := parser.Parse(source, "strict")
	validate.Validate(doc)
	errs := append([]parser.Diag(nil), doc.Errors...)
	sort.SliceStable(errs, func(i, j int) bool { return errs[i].Line < errs[j].Line })
	// Strict-mode validation emits no warnings, so warnings pass through as-is.
	return &Result{Doc: doc, Errors: errs, Warnings: append([]parser.Diag(nil), doc.Warnings...)}
}

// Dump lints source and, when clean, returns its canonical model dump. When the
// document has errors, ok is false and the diagnostics are returned.
func Dump(source string) (out string, diags []parser.Diag, ok bool) {
	res := Lint(source)
	if len(res.Errors) > 0 {
		return "", res.Errors, false
	}
	return dump.Model(model.Build(res.Doc)), nil, true
}
